### PR TITLE
parser: parse decimal literal instead of float literal

### PR DESCRIPTION
**Pull request by coocood**
Improves MySQL compatibility.

### CODE REVIEW

**Comment by coocood** [comment]
@tiancaiamao @hanfei1991 PTAL

File: parser/yy_parser.go [hunk 1]
```diff
@@ -151,6 +152,21 @@ func toInt(l yyLexer, lval *yySymType, str string) int {
 	return intLit
 }
 
+func toDecimal(l yyLexer, lval *yySymType, str string) int {
+	dec := new(mysql.MyDecimal)
```
Comment (ngaut) [inline parser/yy_parser.go]: dec := &mysql.MyDecimal{}

Comment (coocood) [inline parser/yy_parser.go]: Any style guide reference to avoid using `new`?

**Comment by tiancaiamao** [comment]
LGTM

File: parser/yy_parser.go [hunk 2]
```diff
@@ -151,6 +152,21 @@ func toInt(l yyLexer, lval *yySymType, str string) int {
...
+	err := dec.FromString(hack.Slice(str))
+	if err == nil {
+		lval.item = dec
+		return decLit
+	}
+	if terror.ErrorEqual(err, mysql.ErrOverflow) || terror.ErrorEqual(err, mysql.ErrTruncated) {
```
Comment (hanfei1991) [inline parser/yy_parser.go]: add test to cover this case

File: parser/yy_parser.go [hunk 3]
```diff
@@ -151,6 +152,21 @@ func toInt(l yyLexer, lval *yySymType, str string) int {
...
+		return toFloat(l, lval, str)
+	}
+	lval.item = dec
+	l.Errorf("decimal literal: %v", err)
```
Comment (coocood) [inline parser/yy_parser.go]: This is not log, it's the parser error to return.

File: util/types/datum.go [hunk 4]
```diff
@@ -958,7 +958,7 @@ func (d *Datum) convertToMysqlDecimal(target *FieldType) (Datum, error) {
 	case KindString, KindBytes:
 		err = dec.FromString(d.GetBytes())
 	case KindMysqlDecimal:
-		dec = d.GetMysqlDecimal()
+		*dec = *d.GetMysqlDecimal()
```
Comment (hanfei1991) [inline util/types/datum.go]: Why change that?

Comment (coocood) [inline util/types/datum.go]: Create a copy so we do not change the original value.

**Comment by hanfei1991** [comment]
LGTM

**Pull request merged**